File: modules/vision-camera-webrtc-bridge/ios/WebRTCFrameProcessorPlugin.m
#import <VisionCamera/FrameProcessorPlugin.h>
#import <VisionCamera/FrameProcessorPluginRegistry.h>
#import <VisionCamera/Frame.h>

#if __has_include("VisionCameraWebRTCBridge-Swift.h")
#import "VisionCameraWebRTCBridge-Swift.h"
#else
#import <VisionCameraWebRTCBridge/VisionCameraWebRTCBridge-Swift.h>
#endif

@interface WebRTCFrameProcessorPlugin : FrameProcessorPlugin
@end

@implementation WebRTCFrameProcessorPlugin

- (instancetype)initWithProxy:(VisionCameraProxyHolder*)proxy
                  withOptions:(NSDictionary* _Nullable)options {
  self = [super initWithProxy:proxy withOptions:options];
  return self;
}

- (id _Nullable)callback:(Frame*)frame
           withArguments:(NSDictionary* _Nullable)arguments {
  CMSampleBufferRef buffer = frame.buffer;
  if (buffer != NULL) {
    // Forward to the singleton forwarder
    // Swift class is accessed via the generated header
    [VisionCameraFrameForwarder.shared pushFrameWithSampleBuffer:buffer];
  }
  return nil;
}

VISION_EXPORT_FRAME_PROCESSOR(WebRTCFrameProcessorPlugin, forwardToWebRTC)

@end

File: modules/vision-camera-webrtc-bridge/ios/WebRTCFrameProcessorPlugin.swift
import VisionCamera

/**
 * VisionCamera frame processor plugin that forwards each camera frame
 * to `VisionCameraFrameForwarder` for injection into the WebRTC video track.
 *
 * Registered as "forwardToWebRTC" — called from JS via:
 *   `VisionCameraProxy.initFrameProcessorPlugin('forwardToWebRTC')`
 */
@objc(WebRTCFrameProcessorPlugin)
class WebRTCFrameProcessorPlugin: FrameProcessorPlugin {

  override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = nil) {
    super.init(proxy: proxy, options: options)
  }

  override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    let buffer = frame.buffer
    let orientation = frame.orientation
    VisionCameraFrameForwarder.shared.pushFrame(sampleBuffer: buffer, orientation: orientation)
    return 0
  }
}
